controllers/cloudstackcluster_controller.go

```diff
@@ -18,144 +15,136 @@ package controllers
-	}
-
-	// Setup patcher. This ensures modifications to the csCluster copy fetched above are patched into the origin.
-	patchHelper, err := patch.NewHelper(csCluster, r.Client)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
-	defer func() {
-		if err = patchHelper.Patch(ctx, csCluster); err != nil {
-			err = errors.Wrapf(err, "error patching CloudStackCluster %s/%s", csCluster.Namespace, csCluster.Name)
-			retErr = multierror.Append(retErr, err)
-		}
-	}()
+// CloudStackClusterReconciliationRunner is a ReconciliationRunner with extensions specific to CloudStackClusters.
+// The runner does the actual reconciliation.
+type CloudStackClusterReconciliationRunner struct {
+	csCtrlrUtils.ReconciliationRunner
+	Zones                 *infrav1.CloudStackZoneList
+	ReconciliationSubject *infrav1.CloudStackCluster
+	CSUser                cloud.Client
+}
 
-	// Delete Cluster Resources if deletion timestamp present.
-	if !csCluster.DeletionTimestamp.IsZero() {
-		return r.reconcileDelete(ctx, log, csCluster)
-	}
+// CloudStackClusterReconciler is the k8s controller manager's interface to reconcile a CloudStackCluster.
+// This is primarily to adapt to k8s.
+type CloudStackClusterReconciler struct {
+	csCtrlrUtils.ReconcilerBase
+}
 
-	// Reconcile remaining clusters.
-	return r.reconcile(ctx, log, csCluster)
+// Initialize a new CloudStackCluster reconciliation runner with concrete types and initialized member fields.
+func NewCSClusterReconciliationRunner() *CloudStackClusterReconciliationRunner {
+	// Set concrete type and init pointers.
+	runner := &CloudStackClusterReconciliationRunner{ReconciliationSubject: &infrav1.CloudStackCluster{}}
+	runner.Zones = &infrav1.CloudStackZoneList{}
+	// Setup the base runner. Initializes pointers and links reconciliation methods.
+	runner.ReconciliationRunner = csCtrlrUtils.NewRunner(runner, runner.ReconciliationSubject)
+	// For the CloudStackCluster, the ReconciliationSubject is the CSCluster
+	// Have to do after or the setup method will overwrite the link.
+	runner.CSCluster = runner.ReconciliationSubject
+
+	return runner
 }
 
-// Actually reconcile cluster.
-func (r *CloudStackClusterReconciler) reconcile(
-	ctx context.Context,
-	log logr.Logger,
-	csCluster *infrav1.CloudStackCluster,
-) (ctrl.Result, error) {
+// Reconcile is the method k8s will call upon a reconciliation request.
+func (reconciler *CloudStackClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (retRes ctrl.Result, retErr error) {
+	return NewCSClusterReconciliationRunner().
+		UsingBaseReconciler(reconciler.ReconcilerBase).
+		ForRequest(req).
+		WithRequestCtx(ctx).
+		RunBaseReconciliationStages()
+}
 
-	log.V(1).Info("Reconciling CloudStackCluster.", "clusterSpec", csCluster.Spec)
+// Reconcile actually reconciles the CloudStackCluster.
+func (r *CloudStackClusterReconciliationRunner) Reconcile() (res ctrl.Result, reterr error) {
+	return r.RunReconciliationStages(
+		r.RequeueIfMissingBaseCRDs,
```

Comment:
By default the base stages don't exit if a base CRD isn't found.

Deletion needs to be able to proceed even if a base CRD is missing, but most Reconciles need the bases.